packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp

```diff
@@ -0,0 +1,126 @@
+#include <chrono>
+#include "IdleTaskRef.h"
+#include "NativeIdleCallbacks.h"
+
+#ifdef RN_DISABLE_OSS_PLUGIN_HEADER
+#include "Plugins.h"
+#endif
+
+std::shared_ptr<facebook::react::TurboModule> NativeIdleCallbacksModuleProvider(
+    std::shared_ptr<facebook::react::CallInvoker> jsInvoker) {
+  return std::make_shared<facebook::react::NativeIdleCallbacks>(
+      std::move(jsInvoker));
+}
+
+namespace facebook::react {
+
+jsi::Function makeTimeRemainingFunction(
+    jsi::Runtime& runtime,
+    std::shared_ptr<RuntimeScheduler> runtimeScheduler,
+    RuntimeSchedulerTimePoint deadline) {
+  return jsi::Function::createFromHostFunction(
+    runtime,
+    jsi::PropNameID::forAscii(runtime, "timeRemaining"),
+    0,
+    [&](
+      jsi::Runtime& runtime,
+      const jsi::Value& /* unused */,
+      const jsi::Value* /* unused */,
+      size_t /* unused */) noexcept -> jsi::Value {
+        if (runtimeScheduler->getShouldYield()) {
+          return jsi::Value(runtime, 0.0);
+        }
+              
+        auto now = runtimeScheduler->now();
+        double remainingTime = std::chrono::duration_cast<std::chrono::milliseconds>(deadline - now).count();
+        return jsi::Value(runtime, std::max(remainingTime, 0.0));
+    });
+}
+
+NativeIdleCallbacks::NativeIdleCallbacks(std::shared_ptr<CallInvoker> jsInvoker)
+  : NativeIdleCallbacksCxxSpec(std::move(jsInvoker)) {}
+
+CallbackHandle NativeIdleCallbacks::requestIdleCallback(
+  jsi::Runtime& runtime,
+  jsi::Function callback,
+  std::optional<NativeRequestIdleCallbackOptions> options) {
+  using namespace std::chrono_literals;
+
+  auto binding = RuntimeSchedulerBinding::getBinding(runtime);
+  auto runtimeScheduler = binding->getRuntimeScheduler();
+
+  // handle timeout parameter
+  std::optional<double> timeout;
+  if (options.has_value()) {
+    timeout = options.value().timeout;
+  }
+
+  auto userCallback = std::make_shared<jsi::Function>(std::move(callback));
+
+  auto wrappedCallback = jsi::Function::createFromHostFunction(
+    runtime,
+    jsi::PropNameID::forAscii(runtime, "requestIdleCallback$bodyWrapper"),
+    1,
+    [runtimeScheduler, cb = std::move(userCallback)](
+      jsi::Runtime& runtime,
+      const jsi::Value& /* unused */,
+      const jsi::Value* args,
+      size_t /* unused */
+    ) noexcept -> jsi::Value {
+      auto didTimeout = args[0].getBool();
+      // Below is a partial solution and does not comply to WHATWG event-loop standards
+      auto deadline = runtimeScheduler->now() + 50ms;
+
+      jsi::Object idleDeadline {runtime};
+      idleDeadline.setProperty(runtime, "didTimeout", didTimeout);
+      idleDeadline.setProperty(
+        runtime,
+        "timeRemaining",
+        makeTimeRemainingFunction(runtime, runtimeScheduler, deadline)
+      );
+
+      return cb->call(runtime, std::move(idleDeadline));
+    }
+  );
+
+  auto timeoutDuration = std::chrono::duration<double, std::milli>(timeout.value_or(0.0));
```

Comment:
We should use the default param in `scheduleIdleTask` instead of `0` when `timeout` isn't set, because otherwise we'd be using the priority of normal tasks instead of the priority of idle tasks.

Reply:
Hmm, since this function is for idle callbacks only, maybe it would be better to make the `timeout` `max(timeout, timeoutForPriotity(SchedulerPriority::Idle))` so that we don't allow queuing normal tasks with this. WDYT?

Reply:
`timeoutForPriotity(SchedulerPriority::Idle)` is infinite, so that `max` will always give you this value. The moment you specify a timeout you're kinda saying, after this consider this to be a normal task, so I think it should be fine. If anything, we should do something like `expirationTime = now() + timeout + timeoutForPriority(SchedulerPriority::Normal)`, so worst case scenario (timeout = 0) this is scheduled as a normal task. That should be handled in `RuntimeScheduler_Modern` though, and we should still honor the defaults here if the timeout wasn't defined.